crates/bevy_math/src/rects/urect.rs

```diff
@@ -0,0 +1,426 @@
+use crate::UVec2;
+
+/// A rectangle defined by two opposite corners.
+///
+/// The rectangle is axis aligned, and defined by its minimum and maximum coordinates,
+/// stored in `URect::min` and `URect::max`, respectively. The minimum/maximum invariant
+/// must be upheld by the user when directly assigning the fields, otherwise some methods
+/// produce invalid results. It is generally recommended to use one of the constructor
+/// methods instead, which will ensure this invariant is met, unless you already have
+/// the minimum and maximum corners.
+#[repr(C)]
+#[derive(Default, Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub struct URect {
+    /// The minimum corner point of the rect.
+    pub min: UVec2,
+    /// The maximum corner point of the rect.
+    pub max: UVec2,
+}
+
+impl URect {
+    /// Create a new rectangle from two corner points.
+    ///
+    /// The two points do not need to be the minimum and/or maximum corners.
+    /// They only need to be two opposite corners.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::URect;
+    /// let r = URect::new(0, 4, 10, 6); // w=10 h=2
+    /// let r = URect::new(2, 4, 5, 0); // w=3 h=4
+    /// ```
+    #[inline]
+    pub fn new(x0: u32, y0: u32, x1: u32, y1: u32) -> Self {
+        Self::from_corners(UVec2::new(x0, y0), UVec2::new(x1, y1))
+    }
+
+    /// Create a new rectangle from two corner points.
+    ///
+    /// The two points do not need to be the minimum and/or maximum corners.
+    /// They only need to be two opposite corners.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::{URect, UVec2};
+    /// // Unit rect from [0,0] to [1,1]
+    /// let r = URect::from_corners(UVec2::ZERO, UVec2::ONE); // w=1 h=1
+    /// // Same; the points do not need to be ordered
+    /// let r = URect::from_corners(UVec2::ONE, UVec2::ZERO); // w=1 h=1
+    /// ```
+    #[inline]
+    pub fn from_corners(p0: UVec2, p1: UVec2) -> Self {
+        Self {
+            min: p0.min(p1),
+            max: p0.max(p1),
+        }
+    }
+
+    /// Create a new rectangle from its center and size.
+    ///
+    /// # Panics
+    ///
+    /// This method panics if any of the components of the size is negative or if `origin - (size / 2)` results in any negatives.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use bevy_math::{URect, UVec2};
+    /// let r = URect::from_center_size(UVec2::ONE, UVec2::splat(2)); // w=2 h=2
+    /// assert_eq!(r.min, UVec2::splat(0));
+    /// assert_eq!(r.max, UVec2::splat(2));
+    /// ```
+    #[inline]
+    pub fn from_center_size(origin: UVec2, size: UVec2) -> Self {
+        assert!(origin.cmpge(size / 2).all(), "Origin must always be greater than or equal to (size / 2) otherwise you end up subtracting with overflow! Origin was {origin} and size was {size}");
```

Comment:
No, "otherwise the rectangle is undefined". Subtracting with overflow is an implementation detail, the user is not supposed to know how it's related to the construction of the rect.